### PR TITLE
feat: separate products_obsolete MongoDB collection for obsolete products

### LINKED ISSUE #8078: Hide by default the obsolete products withdrawn from the market
### Describe the bug

To avoid confusion for the consumers and from the producers, it would be great to **hide by default the products withdrawn from the market**, from the search results. 

These products could still be visible if scanned or requested by barcode. 

This behaviour would be desirable on both public platform and pro platform. 

Issue created following discussion with @stephanegigandet on [Slack](https://openfoodfacts.slack.com/archives/C02LDQDDD/p1675423981129889) 

 

### To Reproduce

-

### Expected behavior

-

### Screenshots

_No response_

### Additional context

_No response_

### Type of device

Browser

### Browser version

_No response_

### Number of products impacted

_No response_

### Time per product

_No response_

### PR DESCRIPTION
**Pull request by stephanegigandet**
Fixes #8078

This moves obsolete products to a new products_obsolete collection.

?obsolete=1 can be added to URLs to get them to return results from the products_obsolete collection.

### CODE REVIEW

**Comment by stephanegigandet** [comment]
There are a few things that I didn't do yet. e.g. making some scripts or actions run on both products and products_obsolete collections (for instance when we rename the userid of users who want to have their username deleted).

**Review by alexgarel — APPROVED** [review]
Cool

I added some concerns about robustness, I let you decide.

File: scripts/move_obsolete_products_to_products_obsolete_collection.pl [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+	}
+
+	$product_ref = retrieve_product($productid);
+
+	if ((defined $product_ref) and ($productid ne '')) {
+
+		$product_ref->{_id} .= '';
+		$product_ref->{code} .= '';
+		$products_collection->delete_one({"_id" => $product_ref->{_id}});
+		$obsolete_products_collection->replace_one({"_id" => $product_ref->{_id}}, $product_ref, {upsert => 1});
```
[the remark this answers is not in the document]
Comment (stephanegigandet) [reply scripts/move_obsolete_products_to_products_obsolete_collection.pl:80]: Sure, I'll invert the operations.

File: lib/ProductOpener/Products.pm [hunk 2]
```diff
@@ -1021,8 +1039,29 @@ sub store_product ($user_id, $product_ref, $comment) {
...
+			"obsolete status changed",
+			{
+				code => $code,
+				product_id => $product_id,
+				obsolete => $product_ref->{obsolete},
+				was_obsolete => $product_ref->{was_obsolete},
+				previous_products_collection => $previous_products_collection
+			}
+		) if $log->is_debug();
+		$previous_products_collection->delete_one({"_id" => $product_ref->{_id}});
```
[the remark this answers is not in the document]
Comment (stephanegigandet) [reply lib/ProductOpener/Products.pm:1062]: I'll move the delete after the insert.

The case of changed obsolete status and moved to another collection should work as-is (the product will be deleted twice from the previous collection, but that's not a problem).

**Pull request merged**